File: dev/bridge.py
"""Local eval bridge for the BTRoblox DEV build. Stdlib only."""

import json
import queue
import threading
import time
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from urllib.parse import urlparse, parse_qs

HOST = "127.0.0.1"
PORT = 8787

JOB_TIMEOUT = 30.0
POLL_TIMEOUT = 25.0
LOG_LIMIT = 1000

jobs = queue.Queue()
results = {}
results_lock = threading.Lock()

logs = []
logs_lock = threading.Lock()

state = {"next_id": 1, "last_seen": 0.0}
state_lock = threading.Lock()


def next_id():
    with state_lock:
        job_id = state["next_id"]
        state["next_id"] += 1
        return job_id


def touch():
    with state_lock:
        state["last_seen"] = time.time()


def connected():
    with state_lock:
        last = state["last_seen"]
    return last > 0 and time.time() - last < POLL_TIMEOUT + 10


class Handler(BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"

    def log_message(self, fmt, *args):
        pass

    def reply(self, code, obj):
        body = json.dumps(obj).encode("utf8")
        self.send_response(code)
        self.send_header("content-type", "application/json")
        self.send_header("content-length", str(len(body)))
        self.send_header("cache-control", "no-store")
        self.end_headers()
        self.wfile.write(body)

    def body(self):
        length = int(self.headers.get("content-length") or 0)
        return self.rfile.read(length).decode("utf8") if length else ""

    def do_GET(self):
        url = urlparse(self.path)
        args = parse_qs(url.query)

        if url.path == "/poll":
            touch()
            try:
                return self.reply(200, jobs.get(timeout=POLL_TIMEOUT))
            except queue.Empty:
                return self.reply(200, {"idle": True})

        if url.path == "/logs":
            with logs_lock:
                out = list(logs)
                if args.get("clear"):
                    logs.clear()

            level = args.get("level", [None])[0]
            match = args.get("match", [None])[0]

            if level:
                out = [e for e in out if e.get("level") == level]
            if match:
                out = [e for e in out if match in json.dumps(e)]

            return self.reply(200, out[-int(args.get("n", ["50"])[0]):])

        if url.path == "/status":
            return self.reply(200, {
                "connected": connected(),
                "queued": jobs.qsize(),
                "logs": len(logs),
            })

        self.reply(404, {"error": "unknown route"})

    def do_POST(self):
        url = urlparse(self.path)
        args = parse_qs(url.query)
        body = self.body()

        if url.path == "/eval":
            job_id = next_id()
            done = threading.Event()

            with results_lock:
                results[job_id] = {"event": done, "value": None}

            jobs.put({
                "id": job_id,
                "target": args.get("target", ["tab"])[0],
                "code": body,
            })

            if not done.wait(JOB_TIMEOUT):
                with results_lock:
                    results.pop(job_id, None)
                reason = "timed out" if connected() else "extension not connected"
                return self.reply(504, {"ok": False, "error": reason})

            with results_lock:
                slot = results.pop(job_id, None)

            return self.reply(200, slot["value"])

        if url.path == "/result":
            touch()
            try:
                msg = json.loads(body)
            except ValueError:
                return self.reply(400, {"error": "bad json"})

            with results_lock:
                slot = results.get(msg.get("id"))
                if slot:
                    slot["value"] = msg
                    slot["event"].set()

            return self.reply(200, {"ok": True})

        if url.path == "/log":
            touch()
            try:
                entry = json.loads(body)
            except ValueError:
                return self.reply(400, {"error": "bad json"})

            entry["at"] = time.strftime("%H:%M:%S")

            with logs_lock:
                logs.append(entry)
                del logs[:-LOG_LIMIT]

            return self.reply(200, {"ok": True})

        self.reply(404, {"error": "unknown route"})


def main():
    server = ThreadingHTTPServer((HOST, PORT), Handler)
    server.daemon_threads = True
    print(f"btr bridge on http://{HOST}:{PORT}")
    server.serve_forever()


if __name__ == "__main__":
    main()

File: dev/bridge.js
"use strict"

// Local eval bridge for the BTRoblox DEV build. No dependencies.

const http = require("node:http")

const HOST = "127.0.0.1"
const PORT = Number(process.env.BTR_BRIDGE_PORT) || 8787

const JOB_TIMEOUT = 30000
const POLL_TIMEOUT = 25000
const LOG_LIMIT = 1000

const jobs = []
const waiters = []
const pending = new Map()
const logs = []

let nextId = 1
let lastSeen = 0

const connected = () => lastSeen !== 0 && Date.now() - lastSeen < POLL_TIMEOUT + 10000

const reply = (res, code, data) => {
	const body = JSON.stringify(data)

	res.writeHead(code, {
		"content-type": "application/json",
		"content-length": Buffer.byteLength(body),
		"cache-control": "no-store",
		"access-control-allow-origin": "*"
	})

	res.end(body)
}

const readBody = req => new Promise(resolve => {
	let body = ""

	req.setEncoding("utf8")
	req.on("data", chunk => { body += chunk })
	req.on("end", () => resolve(body))
	req.on("error", () => resolve(""))
})

const drop = waiter => {
	const index = waiters.indexOf(waiter)

	if(index !== -1) {
		waiters.splice(index, 1)
		clearTimeout(waiter.timer)
	}
}

const dispatch = () => {
	while(jobs.length && waiters.length) {
		const waiter = waiters.shift()
		clearTimeout(waiter.timer)
		reply(waiter.res, 200, jobs.shift())
	}
}

const routes = {
	"GET /poll"(req, res) {
		lastSeen = Date.now()

		if(jobs.length) { return reply(res, 200, jobs.shift()) }

		const waiter = { res }

		waiter.timer = setTimeout(() => {
			drop(waiter)
			reply(res, 200, { idle: true })
		}, POLL_TIMEOUT)

		res.on("close", () => drop(waiter))
		waiters.push(waiter)
	},

	"POST /eval"(req, res, url, body) {
		const id = nextId++

		const timer = setTimeout(() => {
			pending.delete(id)

			const index = jobs.findIndex(job => job.id === id)
			if(index !== -1) { jobs.splice(index, 1) }

			reply(res, 504, { ok: false, error: connected() ? "timed out" : "extension not connected" })
		}, JOB_TIMEOUT)

		pending.set(id, { res, timer })
		jobs.push({ id, target: url.searchParams.get("target") || "tab", code: body })
		dispatch()
	},

	"POST /result"(req, res, url, body) {
		lastSeen = Date.now()

		let msg

		try { msg = JSON.parse(body) }
		catch(err) { return reply(res, 400, { error: "bad json" }) }

		const entry = pending.get(msg.id)

		if(entry) {
			pending.delete(msg.id)
			clearTimeout(entry.timer)
			reply(entry.res, 200, msg)
		}

		reply(res, 200, { ok: true })
	},

	"POST /log"(req, res, url, body) {
		lastSeen = Date.now()

		try {
			const entry = JSON.parse(body)
			entry.at = new Date().toTimeString().slice(0, 8)

			logs.push(entry)
			while(logs.length > LOG_LIMIT) { logs.shift() }
		} catch(err) { /* ignore malformed */ }

		reply(res, 200, { ok: true })
	},

	"GET /logs"(req, res, url) {
		const level = url.searchParams.get("level")
		const match = url.searchParams.get("match")
		const count = Number(url.searchParams.get("n")) || 50

		let out = logs

		if(level) { out = out.filter(entry => entry.level === level) }
		if(match) { out = out.filter(entry => JSON.stringify(entry).includes(match)) }

		out = out.slice(-count)

		if(url.searchParams.get("clear")) { logs.length = 0 }

		reply(res, 200, out)
	},

	"GET /status"(req, res) {
		reply(res, 200, { connected: connected(), queued: jobs.length, logs: logs.length })
	}
}

const server = http.createServer(async (req, res) => {
	const url = new URL(req.url, `http://${HOST}:${PORT}`)

	if(req.method === "OPTIONS") {
		res.writeHead(204, {
			"access-control-allow-origin": "*",
			"access-control-allow-methods": "GET, POST, OPTIONS",
			"access-control-allow-headers": "content-type",
			"access-control-max-age": "86400"
		})

		return res.end()
	}

	const route = routes[`${req.method} ${url.pathname}`]

	if(!route) { return reply(res, 404, { error: "unknown route", routes: Object.keys(routes) }) }

	const body = req.method === "POST" ? await readBody(req) : ""

	try { route(req, res, url, body) }
	catch(err) {
		if(!res.headersSent) { reply(res, 500, { error: String(err && err.stack || err) }) }
	}
})

server.listen(PORT, HOST, () => console.log(`btr bridge on http://${HOST}:${PORT}`))
